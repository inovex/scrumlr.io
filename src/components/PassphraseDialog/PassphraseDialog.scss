@import "constants/style";

.passphrase-dialog__wrapper {
  display: flex;
  justify-content: center;
  align-items: center;

  min-width: 100vw;
  min-height: 100vh;
}

.passphrase-dialog {
  width: 100%;
  max-width: 420px;

  padding: $padding--default;
  color: $color-black;
}

.passphrase-dialog__form {
  margin-top: 1.5em;
}

.passphrase-dialog__submit-button {
  margin-top: 1em;
}

[theme="dark"] {
  .passphrase-dialog {
    color: $color-white;
  }
}
