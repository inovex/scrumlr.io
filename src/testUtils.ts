export type AssertTypeEqual<T, Expected> =
    T extends Expected
        ? (Expected extends T ? true : never)
        : never;