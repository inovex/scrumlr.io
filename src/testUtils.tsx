import {render, RenderOptions} from "@testing-library/react";
import {I18nextProvider} from "react-i18next";
import {FC, PropsWithChildren, ReactElement} from "react";
import {MemoryRouter} from "react-router";
import {Provider} from "react-redux";
import getTestStore from "utils/test/getTestStore";
import i18n from "./i18nTest";

export type AssertTypeEqual<T, Expected> = T extends Expected ? (Expected extends T ? true : never) : never;

const AllTheProvidersWithoutRouter: FC<PropsWithChildren> = ({children}) => (
  <I18nextProvider i18n={i18n}>
    <Provider store={getTestStore()}>{children}</Provider>
  </I18nextProvider>
);

const AllTheProviders: FC<PropsWithChildren> = ({children}) => (
  <MemoryRouter>
    <AllTheProvidersWithoutRouter>{children}</AllTheProvidersWithoutRouter>
  </MemoryRouter>
);

const customRender = (ui: ReactElement, options?: Omit<RenderOptions & {container: Element}, "wrapper">) => render(ui, {wrapper: AllTheProviders, ...options});
const customRenderWithoutRouter = (ui: ReactElement, options?: Omit<RenderOptions & {container: Element}, "wrapper">) =>
  render(ui, {wrapper: AllTheProvidersWithoutRouter, ...options});

export {customRender as render, customRenderWithoutRouter as renderWithoutRouter};
