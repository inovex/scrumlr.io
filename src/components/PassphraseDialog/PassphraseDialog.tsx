import {FC, useState} from "react";
import "./PassphraseDialog.scss";
import {ScrumlrLogo} from "components/ScrumlrLogo";
import {useTranslation} from "react-i18next";
import {ReactComponent as VisibleIcon} from "assets/icon-visible.svg";
import {ReactComponent as HiddenIcon} from "assets/icon-hidden.svg";
import {TextInputLabel} from "../TextInputLabel";
import {TextInput} from "../TextInput";
import {Button} from "../Button";
import {TextInputAdornment} from "../TextInputAdornment";

export interface PassphraseDialogProps {
  onSubmit: (passphrase: string) => void;
}

export const PassphraseDialog: FC<PassphraseDialogProps> = ({onSubmit}) => {
  const {t} = useTranslation();

  const [passphrase, setPassphrase] = useState<string>("");
  const [visiblePassphrase, setVisiblePassphrase] = useState(false);

  const handleSubmit = async () => {
    await onSubmit(passphrase);
  };

  const togglePassphraseVisibility = () => {
    setVisiblePassphrase(!visiblePassphrase);
  };

  return (
    <div className="passphrase-dialog__wrapper">
      <div className="passphrase-dialog">
        <a href="/" aria-label="Homepage">
          <ScrumlrLogo accentColorClassNames={["accent-color--blue", "accent-color--purple", "accent-color--lilac", "accent-color--pink"]} />
        </a>

        <form className="passphrase-dialog__form" onSubmit={handleSubmit}>
          <TextInputLabel label={t("PassphraseDialog.passphraseInputLabel")} htmlFor="password-dialog__password" />
          <TextInput
            id="password-dialog__password"
            type={visiblePassphrase ? "text" : "password"}
            rightAdornment={
              <TextInputAdornment title={t("PassphraseDialog.togglePassphraseVisibility")} onClick={togglePassphraseVisibility}>
                {visiblePassphrase ? <VisibleIcon /> : <HiddenIcon />}
              </TextInputAdornment>
            }
            value={passphrase}
            onChange={(e) => setPassphrase(e.target.value)}
          />

          <Button type="submit" color="primary" className="passphrase-dialog__submit-button" disabled={!passphrase}>
            {t("PassphraseDialog.submit")}
          </Button>
        </form>
      </div>
    </div>
  );
};

export default PassphraseDialog;
