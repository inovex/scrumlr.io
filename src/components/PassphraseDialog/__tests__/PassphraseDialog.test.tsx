import {fireEvent, waitFor} from "@testing-library/react";
import {render} from "testUtils";
import {PassphraseDialog} from "../PassphraseDialog";

describe("<PassphraseDialog />", () => {
  test("snapshot test", () => {
    const {container} = render(<PassphraseDialog onSubmit={jest.fn()} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  describe("passphrase input", () => {
    test("submit is disabled while passphrase is empty", () => {
      const {container} = render(<PassphraseDialog onSubmit={jest.fn()} />);
      expect(container.querySelector(".passphrase-dialog__submit-button")!.hasAttribute("disabled")).toBeTruthy();
    });

    test("submit is available if passphrase is set", async () => {
      const {container} = render(<PassphraseDialog onSubmit={jest.fn()} />);
      fireEvent.change(container.querySelector("#password-dialog__password")!, {target: {value: "1234"}});

      await waitFor(() => {
        expect(container.querySelector(".passphrase-dialog__submit-button")!.hasAttribute("disabled")).toBeFalsy();
      });
    });
  });

  describe("visibility toggle", () => {
    test("passphrase is not visible in default state", () => {
      const {container} = render(<PassphraseDialog onSubmit={jest.fn()} />);
      expect(container.querySelector("#password-dialog__password")!.getAttribute("type")).toEqual("password");
    });

    test("passphrase is visible on visibility toggle", async () => {
      const {container} = render(<PassphraseDialog onSubmit={jest.fn()} />);
      fireEvent.click(container.querySelector('button[aria-label="Toggle passphrase visibility"]')!);
      await waitFor(() => {
        expect(container.querySelector("#password-dialog__password")!.getAttribute("type")).toEqual("text");
      });
    });
  });

  describe("manual verification hint", () => {
    test("manual verification is visible", () => {
      const {container} = render(<PassphraseDialog onSubmit={jest.fn()} />);
      expect(container.querySelector(".passphrase-dialog__manual-verification")).toBeNull();
    });

    test("manual verification is not visible", () => {
      const {container} = render(<PassphraseDialog onSubmit={jest.fn()} />);
      expect(container.querySelector(".passphrase-dialog__manual-verification")).toBeDefined();
    });
  });
});
