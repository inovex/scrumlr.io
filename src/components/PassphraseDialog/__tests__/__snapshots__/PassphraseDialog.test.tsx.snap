// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PassphraseDialog /> snapshot test 1`] = `
<div
  class="loading-screen"
>
  <div>
    Logo
  </div>
  <div
    class="passphrase-dialog"
  >
    <form
      class="passphrase-dialog__form"
    >
      <label
        class="passphrase-dialog__input-label"
        for="passphrase"
      >
        Enter the passphrase to access this board
      </label>
      <input
        aria-invalid="false"
        id="passphrase"
        type="password"
        value=""
      />
      <button
        aria-label="Toggle passphrase visibility"
        aria-pressed="false"
        type="button"
      >
        Toggle passphrase visibility
      </button>
      <button
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
  <span>
    Collaborative session. By continuing you're accepting our privacy policy.
  </span>
</div>
`;
